### DISCUSSION TITLE
Introduce a new ??? operator for evaluating null/empty/whitespace only strings with a single check, with 30% less code to write than currently and increased code clarity

### DISCUSSION THREAD

**Discussion by pluczak99** [dotnet/csharplang#6714] [Language Ideas]
Frankly, I am baffled about the lack of this feature which should be implemented already many C# versions ago, but if I missed a simple fact that this functionality was already introduced , I apologize in advance for wasting your valuable time.

That last possibility however does not seem very plausible to me, because in new projects at my team, using the latest .NET and C# specification, I still see a lot of "string.IsNullOrWhitespace(str) ?? str2", this is code written by senior developers and it takes 25%-33% more code in implementation than it should. If I was ignorant of this feature existing, I doubt that tens of senior .NET developers working with the latest version of .NET and C# would miss that too, it is unlikely. 

Since "??" operator tests only for nullness (of any nullable type), currently it cannot be used to special case with strings.

I propose introducing a new operator, "???", which works like "??" except it is meant for strings and in addition to checking for nullness of a string, it also checks if the string is empty or whitespaces only and in this case should also return "true".
So the new operator would replace string.IsNullOrEmpty and string.IsNullOrWhitespace, taking 3 characters instead of at minimum 20, without loss of code readibility.

I don't want to get into discussions why a new operator specifically for one type - because that type is quite special unlike other types and the example of use I show below exists in probably every single one sufficiently large project. Personally, having to write string.IsNullOrWhitespace (even with Intellisense) feels like I'm still writing in C# 4.0.

So an example line of code below currently has to look like that:

**`string.IsNullOrWhitespace(str) ?? str2`**

With the new ??? operator, it would look like below:

**`str ??? str2`**

Isn't this more succinct and even more readable than the current solution? Would there be any confusion, even compared to working of "??" operator?

**Comment by yaakov-h** [comment]
I don't see how you can see a lot of `string.IsNullOrWhitespace(str) ?? str2` in your codebase when today this produces:

> error CS0019: Operator '??' cannot be applied to operands of type 'bool' and 'string'

What is your actual use-case?

**Comment by KevinZonda** [comment]
Maybe try extension method?

```cs
public static string IfNullOrWhitespace(this string s, string ifCon)
{
    if (string.IsNullOrWhiteSpace(s)) return ifCon;
    return s;
}
```

```cs
// use
var s = m.IfNullOrWhitespace("Hi");
```

**Reply by TahirAhmadov** [reply]
I would use a different form:
```cs
public static string? NullIfWhiteSpace(this string? This)
{
    if (string.IsNullOrWhiteSpace(s)) return null;
    return This;
}
string res = str.NullIfWhiteSpace() ?? str2;
```

**Comment by BlinD-HuNTeR** [comment]
Instead of adding a new operator to the language, I propose adding the operator `||` to strings, such that `a || b` will only evaluate `b` if `a` is null or empty, exactly as it is in JavaScript.

And if you need to check for whitespace, you can do `a?.Trim() || b`.

This approach requires no change at all to the language, it's just a matter of adding new operator methods to the `System.String` class.

**Reply by DavidArno** [reply]
> ... such that `a || b` will only evaluate `b` if `a` is null or empty, exactly as it is in JavaScript.

JavaScript supports the concept of "truthy" values, ie any value at all can be evaluated to either `true` or `false`. C# has never done this and is extremely unlikely to ever do it in the future. In C#, `""` and `null` are neither `true` nor `false`, in the same way that `"rock"` isn't `1` or `0`.

> This approach requires no change at all to the language, it's just a matter of adding new operator methods to the `System.String` class

C# does not permit adding the `||` operator to classes or structs. It's on [the non-overloadable operator list](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/operator-overloading#non-overloadable-operators). Therefore, it most certainly would require a change to language (and likely a change to .NET itself) to support this.

**Reply by svick** [reply]
@DavidArno [You can effectively overload `||` by overloading `|`, `true` and `false`.](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgWQE8BlYMASygHMACAUwFsAHYI2gXlqnoHcAKAESCAlAG5CpclToAzAPbzO3PkIXzRE3MTKUatAEYBDMMp4DBxsJrx5UmAJz91tZMkMnxdx/yat2bh5gXvi4YPQAxvJgcLQ60jT89gAMtABqRgA2EPQieADeeLTFtKgAzKWYAGxxUnp08sz0YEbA0cj88fW0AM6YMLW6Mr3oIpwAfJXJAHQAkj0AchCZmQDyYACiLGz8fdMZ2bm0APwjtCC9mFolpRX2NQaKmbSNza3RtLJZPfSddcM9MYcSYpOaLZZrTbbIi7fZZHIhG7lSoPJ4vJotNqmcg5P5DfSAia0ACEoPmSxW6y2/lhBwRWgAvkA===) Though I don't think that's likely to happen for `System.String`.

**Comment by theunrepentantgeek** [comment]
> Frankly, I am baffled about the lack of this feature which should be implemented already many C# versions ago,

It's likely not present because it's trivially achievable with features already present in the language. As has been suggested, an extension method is both easy and succinct - and you get to control the name.

However, if a `???` operator were to be introduced, would it use `IsNullOrEmpty()` as the test, or would it use `IsNullOrWhitespace()`?

Whichever one you argue for, there will be concerted opposition who are sure the other option is more correct.

I've worked on systems where we routinely trimmed leading and trailing whitespace, so a definition based on `IsNullOrWhitespace()` would have been useful.

But I've also worked on systems where leading and trailing whitespace was significant, requiring a definition on `IsNullOrEmpty()`.

**Reply by nickpreston24** [reply]
Why not just have strings by default be empty assignments when say, declaring a property or private field?  I've never found a use case where I actually needed `string mystring =null`, because it's silly to even think of having a null in production.  This could be configured as a Code Style, too in Visual Studio / .editorconfig.

**Coders could then focus on detecting whether it's empty, rather than whether it's "null".**

I agree that we shouldn't even have to be talking about it.  So why not kill it for all time, instead of the ! operator and ?? or ??? or Nullable<string> and all the other 3-headed monsters?

**Reply by CyrusNajmabadi** [reply]
How would you tell the difference from uninitialized versus empty?

**Reply by iam3yal** [reply]
@MikePreston17 There are places where an empty value can actually have meaning so it's not silly at all and if you think that it's silly to "even think of having null in production" then you should probably reconsider your stance because fact is the concept of nullability is important and you might be confusing two aspects of the value the concept of something being null as @CyrusNajmabadi pointed above with its implementation in many languages that unfortunately results in the "billion-dollar mistake" but fortunately today there are ways to express that in the language to avoid the mistake to a degree.

**Comment by DavidArno** [comment]
Pattern matching offers an elegant alternative to the `string.IsNullOrEmpty()` imperative approach:
```cs
str switch {
    {} s and not "" => s,
    _ => str2
};
```

And for the `string.IsNullOrWhiteSpace()` crowd:
```cs
str switch {
    {} s when s.Trim() != "" => s,
    _ => str2
};
```

At the moment though, there's no way to test for whitespace-only string without using a `when` clause. That will become solvable if/when active patterns land. To be honest, I'd much rather the team spent time on a generalised, active patterns feature, rather than yet another obscure edge-case operator that will only satisfy the needs of 50% of that edge-case crowd.

**Reply by iam3yal** [reply]
Why would you use this over `string.IsNullOrEmpty()`? is this really more readable? pattern matching as a concept is declarative but here you're using it imperatively, yes, the compiler generates the code but where is the benefit? I'd say it's a bad example. 😄

**Reply by DavidArno** [reply]
@eyalalonn,

That is the problem of simplified examples: they all too often simply too far and miss the point. The benefit comes from the fact that rarely does one just test for null, empty and whitespace and do no other processing of the string. Using pattern matching lets it all be handled in one switch, rather than nesting if's and switches. So a better example might be:

```cs
foreach (var str in strList)
{
    switch (str) 
    {
        case not {} or "":
            break; //just ignore empty strings
        case var s when s.Trim() == "":
            LogAndIgnoreWhitespaceOnlyString();
            break;
        case ['_', ..] command:
            HandleCommand(command);
            break;
        case ...
    }
}
```
The point being, these situations can be pattern-matched these days, so what need is there for obscure operators?

Not sure if my p[oint is any clearer though or if I've just got an obsession with pattern matching *everything* 😃

**Reply by HaloFour** [reply]
Which would only get better with active patterns.

**Comment by iam3yal** [comment]
@pluczak99 
> Personally, having to write string.IsNullOrWhitespace (even with Intellisense) feels like I'm still writing in C# 4.0.

Really now? and having an operator like `???` is better than something that everyone understand just by looking at the name of the method? maybe you feel like writing `C# 4.0` that btw doesn't make sense because methods were introduced with the first version of the language and using alternative syntax just because you feel like methods isn't good enough representative and you're bored with it doesn't justify the costs.

As noted by @yaakov-h the following doesn't even compile `string.IsNullOrWhitespace(str) ?? str2` so you probably mean something like `string.IsNullOrWhitespace(str) ? str2 : str` and if you find it tedious/repetitive or that you do it more than you like the problem might not be with a specific method in your codebase but with the flow and abstractions in the codebase.
